### PR TITLE
[Interop] Implement GAMMA session drain test

### PR DESCRIPTION
**Pull request by gnossen**
Depends on https://github.com/grpc/grpc/pull/34518

### CODE REVIEW

File: tools/run_tests/xds_k8s_test_driver/framework/rpc/grpc_testing.py [hunk 1]
```diff
@@ -159,6 +161,25 @@ def set_not_serving(self):
             rpc="SetNotServing", req=empty_pb2.Empty(), log_level=logging.INFO
         )
 
+    def send_hook_request_start_server(self):
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/rpc/grpc_testing.py:164]: `HookService` [is not a part of](https://github.com/grpc/grpc/blob/e76ac4ab07d936a857339fb1bd4eeb621bd3fba6/src/proto/grpc/testing/test.proto#L95-L110) `XdsUpdateHealthService`, so it shouldn't be in `XdsUpdateHealthServiceClient`.

We should create another class called `HookServiceClient`.

File: tools/run_tests/xds_k8s_test_driver/framework/rpc/grpc_testing.py [hunk 2]
```diff
@@ -39,6 +39,8 @@
 MethodStats = messages_pb2.LoadBalancerAccumulatedStatsResponse.MethodStats
 RpcsByPeer = messages_pb2.LoadBalancerStatsResponse.RpcsByPeer
 
+_HOOK_SERVER_PORT = 8000
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/rpc/grpc_testing.py:42]: Is this hardcoded in the implementation as well?

File: tools/run_tests/xds_k8s_test_driver/tests/gamma/affinity_test.py [hunk 3]
```diff
@@ -166,5 +170,121 @@ def test_session_affinity_policy_with_service_target(self):
...
+        with self.subTest("04_start_test_client"):
+            test_client: _XdsTestClient = self.startTestClient(test_servers[0])
+
+        with self.subTest("05_send_first_RPC_and_retrieve_cookie"):
+            (
+                cookie,
+                chosen_server,
+            ) = session_affinity_util.assert_eventually_retrieve_cookie_and_server(
+                self, test_client, test_servers
+            )
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/tests/gamma/affinity_test.py:218]: nit:  hard to read - looks too much like lisp :)

**Review by sergiitk — CHANGES_REQUESTED** [review]
initial comments

File: tools/run_tests/xds_k8s_test_driver/framework/rpc/grpc_testing.py (hunk 2 above)
Comment (gnossen) [reply tools/run_tests/xds_k8s_test_driver/framework/rpc/grpc_testing.py:42]: Hardcoded in the `grpcurl` invocation in the server yaml file. I'm working with Eugene on an improved hook server API that doesn't require port management at all.

File: tools/run_tests/xds_k8s_test_driver/framework/rpc/grpc_testing.py (hunk 1 above)
Comment (gnossen) [reply tools/run_tests/xds_k8s_test_driver/framework/rpc/grpc_testing.py:164]: [`SendHookRequest` _is_ and that's what's being invoked here](https://github.com/grpc/grpc/blob/a17f08b49d7e46a2a102c84f57086fd1e6312113/src/proto/grpc/testing/test.proto#L109). There are two different RPCs here:

- `HookService.Hook`, which you pointed out here and which is called by `grpcurl`
- `XdsUpdateHelathService.SendHookRequest`, which controls a server running `HookService.Hook` and which is the actual RPC method referenced here.

File: tools/run_tests/xds_k8s_test_driver/tests/gamma/affinity_test.py (hunk 3 above)
Comment (gnossen) [reply tools/run_tests/xds_k8s_test_driver/tests/gamma/affinity_test.py:218]: Hey, I love Lisp! Regardless, this wasn't my decision, it was Black's. :shrug:

File: tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py [hunk 4]
```diff
@@ -954,6 +951,7 @@ def port_forward_pod(
             self._api.context,
             self.name,
             f"pod/{pod.metadata.name}",
+            pod.metadata.name,
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py:954]: let's use `pod.metadata.uid` instead:

> uid: a unique in time and space value (typically an RFC 4122 generated identifier, see [the identifiers docs](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/)) used to distinguish between objects with the same name that have been deleted and recreated
> — https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata

**Comment by sergiitk** [comment]
Porting to the framework in the new repo: https://github.com/grpc/psm-interop/pull/34

**Comment by sergiitk** [comment]
Fully ported to grpc/psm-interop#34 now, all future work will be done there.

File: tools/run_tests/xds_k8s_test_driver/framework/rpc/grpc_testing.py (hunk 1 above)
Comment (sergiitk) [reply tools/run_tests/xds_k8s_test_driver/framework/rpc/grpc_testing.py:164]: Conversation moved to https://github.com/grpc/psm-interop/pull/34#discussion_r1484906928.

File: tools/run_tests/xds_k8s_test_driver/framework/rpc/grpc_testing.py (hunk 2 above)
Comment (sergiitk) [reply tools/run_tests/xds_k8s_test_driver/framework/rpc/grpc_testing.py:42]: Conversation moved to https://github.com/grpc/psm-interop/pull/34#discussion_r1484908031.

File: tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py (hunk 4 above)
Comment (sergiitk) [reply tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py:954]: Conversation moved to https://github.com/grpc/psm-interop/pull/34#discussion_r1484907499.

**Pull request closed without merging**